### ISSUE TITLE
Using `+` in a document `_id`

### ISSUE THREAD

**Issue by samwarfield** [IBM-Swift/Kitura-CouchDB#104]
I have a `+` character in some of my `_id`'s

I am able to create a document with `+` as part of the `_id`. Then if I try to retrieve by the same id, it is not found.

```
let id = "abc+1"
let document = Document(id: id)

database.create(document) { response, error in
    // document created with ok (can see it using the Fauxton UI)

    // Try to get it.
    database.retrieve(id) { document, error in
        // error.statusCode will be .notFound here
    }
}
```

I noticed in `HTTP.swift`
```
/// A set of characters that are valid in requests.
private static let allowedCharacterSet =  NSCharacterSet(charactersIn:"\"#%/<>?@\\^`{|} ").inverted
```

Is the right thing to do add it to here? of it there a reason we don't have `+` in the allowed character set?

**Comment by pushkarnk** [comment]
@samwarfield If you are referring to [this](https://github.com/IBM-Swift/Kitura-net/blob/master/Sources/KituraNet/HTTP/HTTP.swift#L158), it is the characters allowed in a URL. Characters outside this will be percent encoded. This must relate to document `_id`s if they are sent as a part of the URL, in which case `+` will be percent encoded.

Can you tell us more about the problem(s) you are facing? Your use case, failure symptom etc. Thanks.

**Comment by samwarfield** [comment]
I am referring to that...

When I query by _id, the query gets encoded. I have a + character in the _id.

My problem is, the + is not percent encoded.

**Comment by samwarfield** [comment]
I edited original post with more details

**Comment by pushkarnk** [comment]
Thanks @samwarfield

**Comment by pushkarnk** [comment]
This issue can be reproduced in [this test](https://github.com/IBM-Swift/Kitura-CouchDB/blob/master/Tests/CouchDBTests/DocumentCrudTests.swift#L34) by adding a `+` to the id string. The reported error (seen both with Kitura-net and Kitura-NIO) is:

```
Kitura-CouchDB/Tests/CouchDBTests/DocumentCrudTests.swift:82: error: DocumentCrudTests.testCrudTest : failed - Error in reading document Optional(404) Optional("Error: not_found, missing")
```

**Comment by pushkarnk** [comment]
Though `+` is permissible in the URL path and in CouchDB document ids, it looks like the couchdb agent expects it to be percent-encoded. Here's an example:

`curl http://127.0.0.1:5984/kitura_test_db/_all_docs`
```
{"total_rows":3,"offset":0,"rows":[
{"id":"123+456","key":"123+456","value":{"rev":"1-37a2819c6e288ee92df85209a6c4bce0"}},
{"id":"654321","key":"654321","value":{"rev":"1-92c78e73f7a06c12e99f76560fc8a495"}},
{"id":"abcdef","key":"abcdef","value":{"rev":"1-390dfc9146f9bf1f3b6a86f8ba9832c2"}}
]}
```

`curl http://127.0.0.1:5984/kitura_test_db/123+456`
```
{"error":"not_found","reason":"missing"}
```

`curl http://127.0.0.1:5984/kitura_test_db/123%2B456`
```
{"_id":"123+456","_rev":"1-37a2819c6e288ee92df85209a6c4bce0","value":"value1","created_at":"Tue Aug 28 21:16:23 +0000 2012","truncated":false,"favorited":false}
```

**Comment by djones6** [comment]
@samwarfield this is now fixed in the [3.2.0 release](https://github.com/IBM-Swift/Kitura-CouchDB/releases/tag/3.2.0).
